cmd/private-org-sync/main.go

```diff
@@ -398,6 +392,67 @@ func (g gitSyncer) mirror(repoDir string, src, dst location) error {
 	return nil
 }
 
+func addGitRemote(logger *logrus.Entry, git gitFunc, org, repo, repoDir, remoteName string) error {
+	remoteSetupLogger := logger.WithField("remote-name", remoteName)
+	remoteSetupLogger.Debug("Remote does not exist, setting up")
+
+	srcUrl, err := url.Parse(fmt.Sprintf("https://github.com/%s/%s", org, repo))
+	if err != nil {
+		remoteSetupLogger.WithError(err).Error("Failed to construct URL for the source remote")
+		return fmt.Errorf("failed to construct URL for the source remote")
+	}
+
+	remoteSetupLogger = remoteSetupLogger.WithField("remote-url", srcUrl.String())
+	remoteSetupLogger.Debug("Adding remote")
+
+	if _, exitCode, err := git(logger, repoDir, "remote", "add", remoteName, srcUrl.String()); err != nil || exitCode != 0 {
+		remoteSetupLogger.WithField("exit-code", exitCode).WithError(err).Error("Failed to set up source remote")
+		return fmt.Errorf("failed to set up source remote")
+	}
+
+	return nil
+}
+
+func mergeRemotesAndPush(logger *logrus.Entry, git gitFunc, repoDir, srcRemote, destRemote, org, repo, branch, destUrl string, confirm bool) error {
+	if err := addGitRemote(logger, git, org, repo, repoDir, destRemote); err != nil {
+		return err
+	}
+
+	sourceBranch := fmt.Sprintf("%s/%s", srcRemote, branch)
+	if _, _, err := git(logger, repoDir, []string{"merge", sourceBranch}...); err != nil {
```

Comment:
This looks very convoluted... we basically need to do three operations: fetch the destination (and only the destination branch), merge the source branch (which we already have from the earlier operations) into it, and push it to the destination... There's no need to merge twice (a `checkout` suffices for what the first merge achiveves here). There's also no need to create a local named branch because you can use the `FETCH_HEAD` and `HEAD` identifiers.

Reply:
Please note that this basically says the same like https://github.com/openshift/ci-tools/pull/825#discussion_r429241984 does, it is just more easily visible on the testcase.